components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1664,6 +1802,47 @@ where
                 self.finish_unsafe_recovery_wait_apply();
             }
         }
+        if self.fsm.unsafe_recovery_wait_failed_voters_demotion_apply {
+            if self.fsm.unsafe_recovery_num_tasks.is_none() {
+                panic!(
+                    "{:?} is waiting for failed peers' demotion to apply, but the number of total
+		    recovery tasks is not set",
+                    self.fsm.peer.region()
+                );
+            }
+            self.fsm.unsafe_recovery_wait_failed_voters_demotion_apply = false;
+            let num_tasks = self.fsm.unsafe_recovery_num_tasks.take().unwrap();
+
+            // Exit joint state and wait for it to be applied.
```

Comment:
how do you ensure the enter joint is applied, maybe it's the apply result of other command.

Reply:
you can leverage the `CallBack` passed when proposing the command to inform it's applied